src/main/scala/mesosphere/marathon/core/heartbeat/HeartbeatActor.scala

```diff
@@ -0,0 +1,97 @@
+package mesosphere.marathon.core.heartbeat
+
+import akka.actor._
+import scala.concurrent.duration._
+
+// HeartbeatActor monitors the heartbeat of some process and executes handlers for various conditions.
+// If an expected heartbeat is missed then execute an `onSkip` handler.
+// If X number of subsequent heartbeats are missed then execute an `onFailure` handler and become inactive.
+// Upon creation the actor is in an inactive state and must be sent a MessageActivate message to activate.
+// Once activated the actor will monitor for MessagePulse messages (these are the heartbeats).
+// The actor may be forcefully deactivated by sending it an MessageDeactivate message.
+class HeartbeatActor(config: Heartbeat.Config) extends LoggingFSM[HeartbeatInternal.State, HeartbeatInternal.Data] {
+  import Heartbeat._
+  import HeartbeatInternal._
+
+  startWith(StateInactive, DataNone)
+
+  when(StateInactive) {
+    case Event(MessageActivate(reactor, token), DataNone) =>
+      log.debug("heartbeat activated")
+      goto(StateActive) using DataActive(reactor, token)
+    case _ =>
+      stay // swallow all other event types
+  }
+
+  when(StateActive, stateTimeout = config.heartbeatTimeout) {
+    case Event(MessagePulse, data: DataActive) =>
+      stay using data.copy(missed = 0)
+
+    case Event(StateTimeout, data: DataActive) =>
+      if (data.missed + 1 >= config.missedHeartbeatsThreshold) {
+        data.reactor.onFailure
```

Comment:
Ah I didn't follow the lines then. Seemed appropriate to have them where the HeartbeatActor _decides_ to call `onFailure`. Not generally fond of redundant logs, although, in this case, it would just log _hey I'm calling onFailure_, then _hey, onFailure was called_ and then abdicate, so not really polluting the logs.

Long story short: I have a slight inclination towards logging here, and I wouldn't mind redundancy in this case.